### PR TITLE
Adds ability to apply transactions to StorageEngine API

### PR DESCRIPTION
**Pull request by tinwelint**
Contains a rework of the commit process where the basic unit to commit is a batch of transactions instead of a single transaction. Some work is temporary with the knowledge that ValidatedIndexUpdates are going to be removed in the near future.

### CODE REVIEW

**Comment by chrisvest** [comment]
Nice javadocs

**Comment by davidegrohmann** [comment]
@tinwelint needs rebase

File: community/kernel/src/main/java/org/neo4j/kernel/api/txstate/TxStateVisitor.java [hunk 1]
```diff
@@ -38,301 +38,240 @@
  */
 public interface TxStateVisitor
 {
-    void visitCreatedNode( long id );
+    default void visitCreatedNode( long id ) {}
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/api/txstate/TxStateVisitor.java]: I am unsure this is a good idea. I would like the compiler to report errors if there are missing method in the class implementing this interface.

Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/api/txstate/TxStateVisitor.java]: Oh OK, you have a point there... I'll do the normal Adapter/Delegator thing instead.

**Comment by MishaDemianenko** [comment]
@tinwelint in couple of interfaces we expand `IOException` into `Exception` (for example in `ResponsePacker`, but it's all over the code ), IOException was already too generic, but now it become even bigger, as soon as you see `Exception` in a signature or part of some value object only one thing that you can with it - re-throw or ignore

**Comment by tinwelint** [comment]
@MishaDemianenko yes that was a slight creep. Some implementations of those methods now needed to throw f.ex. `TransactionFailureException` and that's why I expanded it. I'm open to suggestions there. You can also look at it like this: at this point in the life of a transaction, if anything goes wrong when applying, then whatever happens should be treated just as seriously, i.e. kernel panic.

**Comment by tinwelint** [comment]
@MishaDemianenko green again and most issues we talked about solved, look at the last commit

**Pull request merged**